proposals/0304-structured-concurrency.md

```diff
@@ -717,119 +765,238 @@ The isCancelled function is implemented as:
+
+Note also that the `withThrowingTaskGroup` uses a `ThrowingTaskGroup<ChildTaskResult, Error>`, however specifying the type of that error is not possible. This is because this Failure parameter on the `ThrowingTaskGroup` in only used as future-proof API in case Swift were to gain typed throwing at some point in time. This design makes no promises nor does it assume typed throws are actually going to happen though.
 
 A task group _guarantees_ that it will `await` all tasks that were added to it before it returns.
 
 This waiting can be performed either: 
 - by the code within the task group itself (e.g., using `next()` repeatedly until it returns `nil`, described below), or
 - implicitly in the task group itself when returning from the `body`.
 
-By default, the task group will schedule child tasks added to the group on the default global concurrent executor. The `startingChildTasksOn` argument can
-be provided to override this behavior.
+By default, the task group will schedule child tasks added to the group on the default global concurrent executor. In the future is is likely that it will be possible to customize the executor tasks are started on with an optional executor parameter to `spawn`, e.g. like `startingChildTasksOn:`.
 
-##### Adding tasks to a group
+##### Spawning TaskGroup child tasks
 
-Within the `body` function, tasks may be added dynamically with the `add` operation. Each task produces a value of the same type (the `ResultType` generic parameter):
+Within the `body` function, tasks may be added dynamically with the spawn operation. Each task produces a value of the same type (the `ResultType` generic parameter):
 
 ```swift
-extension Task.Group { 
+extension TaskGroup {
   /// Add a child task to the group.
   ///
-  /// Returns true if the task was successfully added, false otherwise.
-  mutating func add(
+  /// Returns true if the task was successfully spawned, false otherwise.
+  /// A cancelled group will reject any further additions of tasks into it.
+  mutating func spawn(
       overridingPriority: Priority? = nil,
-      operation: @concurrent @escaping () async throws -> TaskResult
-  ) async -> Bool
+      operation: @concurrent @escaping () async -> TaskResult
+  ) -> Self.Spawned
+  
+  public struct Spawned: Sendable {
+    /// Returns `true` if the task was successfully spawned in the task group,
+    /// `false` otherwise which means that the group was already cancelled and
+    /// refused to accept spawn a new child task.
+    public var successfully: Bool 
+
+    /// Task handle for the spawned task group child task,
+    /// or `nil` if it was not spawned successfully.
+    public let handle: Task.Handle<ChildTaskResult, Never>?
+  }
```

Comment:
Another alternative design for this is with throwing `CancellationError` however then ONLY the `withThrowingGroup` could do so, and we'd need some new name for `trySpawn` (try spawn is terrible since it becomes `try group.trySpawn { }`)

```swift
withTaskGroup(of: Int.self) { group in 
  // always spawns:
  group.spawn {}
}

withThrowingTaskGroup(of: Int.self) { group in 
  // always spawns:
  group.spawn {} 

  // spawns, unless the group/owning-task were cancelled, then it throws:
  try group.spawnUnlessCancelled {} 
}
```

So that'd be another option; and the spawn would be Void returning OR returning `Task.Handle` and `Task.Handle?` in the "may fail" case.

FYI @airspeedswift for when you have the time to review; plenty of various tradeoffs here, I hope I explained them well though, it is mostly: 
- ability to not spawn in a cancelled group
- ability to always spawn (as async lets would do, if we express them as "implicitly in a group" which I think we should, but I did not manage to do this with my beginner SIL skill...)
- ability to get a handle for a specific spawned task

The Spawned addressess all in one API call though one has to search for it. The throwing versions are very annoying, and can only be available in `withThrowingTaskGroup`. So there's some tradeoffs to make here I'd need input on. Thanks in advance!

Reply:
You can still `spawnUnlessCancelled` inside a non-throwing group, you'd just need to handle the `CancellationError` locally and not let it propagate out of the group block.

Reply:
If the error handling ceremony is too onerous, another option is to have `spawnUnlessCancelled` return a `Bool` like `add` was previously proposed to, but remove the `@discardableResult` from the declaration, so that you have to look at the result. I don't think it needs to return more information than a `Bool`, though.

Reply:
Great points, thank you! I'll adjust this right away so it is part of the review 👍